### PR TITLE
Tutorial for issue #35, initialize model with dirty image

### PR DESCRIPTION
**Pull request by RCF42**
This is a WIP slightly, I believe I've updated all the documentation and everything else correctly as well. 
Tutorial goes over the steps to get the model optimized based off the dirty image and then saving/loading the model
. Just slightly unsure about certain phrasings and want to make sure my loss function is right. Also need to embed links.

### CODE REVIEW

File: docs/tutorials/initializedirtyimage.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+gridder = gridding.Gridder(
+    coords=coords,
+    uu=uu,
+    vv=vv,
+    weight=weight,
+    data_re=data_re,
+    data_im=data_im,
+)
+
+# export to PyTorch dataset
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:79]: I cloned your fork and tried to build the docs on your branch, but got an error here. Your tutorial is written for the new syntax that Brianna created (which is good), but I think your fork still needs to incorporate the upstream changes from MPoL-dev/main. I think you can do this easily using the "Fetch Upstream" changes button on the main page of your forked repo. I'm kind of confused how this builds correctly for you locally when you run `make html`? It seems like it builds correctly in the pull request because those changes are being applied on top of the new syntax changes already in MPoL-dev/main.

File: docs/tutorials/initializedirtyimage.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+# + nbsphinx="hidden"
+# %run notebook_setup
+# -
+
+# # Not completely done, still have to do some links do some fleshing out, just want to get it here to make sure there isn't anything critically wrong.
+#
+# ## Initializing Model with the Dirty Image
+#
+# In this tutorial we will be going over how to optimize the model based off of the dirty image. It has a low loss value (maximal likelihood) and so this allows the optimization to be much quicker. The problem with the dirty image is that certain model paramterizations enforce image positivity while the dirty image usually contains regions of negative flux. So in this tutorial we will be optimizing the model to be similar to the dirty image (phrasing could be better here). We will also be going over saving and loading the model. After saving and loading it, it can be then optimized against the visibility data to complete it (though this will not be done in this tutorial). We will be using the dataset of the ALMA logo first used in the Gridding and Diagnostic Images tutorial.
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:27]: I would back up a little bit and provide more context. 

* Our RML algorithm is written as an optimization loop (see previous optimization tutorial)
* The number of iterations required to reach the optimal point will be fewer if we have a good starting point for the model parameters
* Currently we're just starting from a flat (uniform) image. The dirty image is a better starting point, since we've already established that it maximizes the likelihood function (if not priors)
* We've implemented an RML model that does not permit negative flux values. This is a good thing for astronomical image reconstruction, because real images don't have negative flux. 
* However, this presents a problem for using the dirty image as a starting point, since it contains negative values.
* This tutorial shows how to initialize the model parameters to reproduce the dirty image as closely as possible, under the constraint that all pixels must be positive

File: docs/tutorials/initializedirtyimage.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    data_re=data_re,
+    data_im=data_im,
+)
+
+# export to PyTorch dataset
+dset = gridder.to_pytorch_dataset()
+# -
+
+# Show the dirty image
+img, beam = gridder.get_dirty_image(weighting="briggs", robust=0.0, unit="Jy/beam")
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:84]: It would be useful to accentuate the negative values in the dirty image using a diverging colormap: https://matplotlib.org/stable/tutorials/colors/colormaps.html

File: docs/tutorials/initializedirtyimage.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+# The results are seen here with the state_dict() having changed and the image now more closely resembling the dirty image, though the dirty image contains negative values which this does not so they can never be completely the same.
+
+fig, ax = plt.subplots(nrows=1)
+im = ax.imshow(
+    np.squeeze(rml.icube.sky_cube.detach().numpy()),
+    origin="lower",
+    interpolation="none",
+    extent=rml.icube.coords.img_ext,
+)
+plt.colorbar(im)
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:164]: We can use the same colorbar as the dirty image to show that there are no negative values in this representation

File: docs/tutorials/initializedirtyimage.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+
+# ### Loss Function
+
+# The loss function that will be used to optimize the initial part of the image is the pixel-to-pixel L2 norm (also known as the Euclidian Norm). It calculates the loss based off of the image-plane distance between the dirty image and the state of the ImageCube in order to make the state of the ImageCube closer to the dirty image.
+#
+# The L2 Norm (Euclidian Norm) is based off of the euclidian distance function and it takes the form of: $ L_{oss} = \sqrt{\sum_{i=1}^{N} (y_i - x_i)^2}  $
+
+
+def l2norm(sky, dirt):
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:112]: This loss function works, but I think it would be simpler to just use `loss_fn = torch.nn.MSELoss()`

File: docs/tutorials/initializedirtyimage.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    loss_tracker.append(loss.item())
+    loss.backward()
+    optimizer.step()
+# torch.save(rml, "dirty_image_model.pt")
+torch.save(rml.state_dict(), "dirty_image_model.pt")
+# -
+
+# We also can save the state of the model or the entire model itself. The torch.save(model.state_dict(), PATH) will just save the state.dict() which are the model's learned parameters. The alternative will save the entire model.
+# https://pytorch.org/tutorials/beginner/saving_loading_models.html
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:146]: I think we want to steer the user towards saving `model.state_dict` over the entire model, since state dict is a much simpler object to understand.

File: docs/tutorials/initializedirtyimage.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+plt.colorbar(im)
+
+rml.state_dict()
+
+# ### Loading the Model
+#
+# To demonstrate the saving and loading of the model here we will be resetting the parameters of the model and then reloading them.
+
+rml = precomposed.SimpleNet(coords=coords)
+rml.state_dict()  # parameters of the model
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:173]: These are the *uninitialized* parameters of the model, or with the default initializations

File: docs/tutorials/initializedirtyimage.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+# To continue and optimize our model against the visibility data the gridded visibilitied need to have uniform weighting
+
+gridder._grid_visibilities(weighting="uniform")
+img, beam = gridder.get_dirty_image(weighting="uniform")
+dataset = gridder.to_pytorch_dataset()  # creating a new dataset
+
+# ### Conclusion
+#
+# This tutorial shows how to work towards a better starting point for RML optimization by using an image with maximal likelihood but RML optimization does not need to start from here and can start from a blank image. This is unlike a deconvolution algorithm such as the CASA tclean. The training loop done here is just the set up and the model/image still needs further regularization and optimization against the visibility data.
```
Comment (iancze) [inline docs/tutorials/initializedirtyimage.py:188]: You can link to the optimization and cross validation tutorials for more information

**Review by iancze — CHANGES_REQUESTED** [review]
This looks like a very nice start. I have some suggestions for colorbar plots. 

Also, it would be helpful if you could pull upstream changes to your local fork. I can't seem to build the documentation locally, so I'm just editing things based off of what I see in the source code here.

File: docs/tutorials/initializedirtyimage.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+fig, ax = plt.subplots(ncols=1)
+snp = ax.imshow(np.squeeze(img), **kw)
+ax.set_title("image")
+ax.set_xlabel(r"$\Delta \alpha \cos \delta$ [${}^{\prime\prime}$]")
+ax.set_ylabel(r"$\Delta \delta$ [${}^{\prime\prime}$]")
+plt.colorbar(snp)
+
+# ### Model and Optimization Setup
+#
+# Here we will be setting up the optimizer and the image model (RML). If this is unfamiliar please reference the Optimization tutorial. The initial parameters of the model are also displayed and can be contrasted with them after the optimization loop. }
```
Comment (briannazawadzki) [inline docs/tutorials/initializedirtyimage.py:95]: I think it would be nice to include a link to the optimization tutorial here. Also, this may be a bit nitpicky, but I think you should consider rephrasing the tutorial to use more direct language, e.g. "Here we will be setting up the optimizer" --> "Here we set up the optimizer"

File: docs/tutorials/initializedirtyimage.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+rml.state_dict()
+
+# ### Loading the Model
+#
+# To demonstrate the saving and loading of the model here we will be resetting the parameters of the model and then reloading them.
+
+rml = precomposed.SimpleNet(coords=coords)
+rml.state_dict()  # parameters of the model
+
+# Here you can clearly see the state_dict() returning to its original state from the beginning of the tutorial and then once we reload it, it reverts to the values it was when we saved it.
```
Comment (briannazawadzki) [inline docs/tutorials/initializedirtyimage.py:175]: Can you rephrase this for clarity? I know what you mean, but I think it may be a bit confusing for a reader who is new to this.

File: docs/tutorials/initializedirtyimage.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+rml.load_state_dict(torch.load("dirty_image_model.pt"))
+rml.state_dict()  # parameters of the model
+
+# To continue and optimize our model against the visibility data the gridded visibilitied need to have uniform weighting
+
+gridder._grid_visibilities(weighting="uniform")
+img, beam = gridder.get_dirty_image(weighting="uniform")
+dataset = gridder.to_pytorch_dataset()  # creating a new dataset
```
Comment (briannazawadzki) [inline docs/tutorials/initializedirtyimage.py:184]: I'm not sure what purpose these three lines serve. You already called `get_dirty_image` and `to_pytorch_dataset` earlier in the tutorial. I don't think you should call `_grid_visibilities` in the tutorial at all now since the user is no longer meant to access that function directly. Using `get_dirty_image` and/or `to_pytorch_dataset` should already grid the visibilities in the way the user needs.

File: docs/tutorials/initializedirtyimage.py (hunk 1 above)
Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:79]: I made sure I had done fetch origin and everything before doing the pull request (or at least I thought I did) and when I do make html the documentation was all building properly for me. I'll try and see what's up with that.

Comment (iancze) [reply docs/tutorials/initializedirtyimage.py:79]: I think fetch origin will just update your local repository with the changes to RCF42/MPoL, which should be none, since you were the only one making changes. What you want to do is update RCF42/MPoL with the new changes that are in MPoL-dev/MPoL. I think you can do this using the "Fetch Upstream" button I mentioned. See also here: https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/merging-an-upstream-repository-into-your-fork

Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:79]: Ah okay, my bad

Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:77]: So I've done the fetch upstream and made sure that my branch is up to date with the overall main but I'm facing a problem that when I try and build the documentation it will make just about everything fine but when it comes to the tutorials it gives me:
 `E:\robert\Documents\GitHub\MPoL\docs\index.rst:33: WARNING: toctree contains reference to nonexisting document 'tutorials/gridder'`
for every single tutorial and nothing relating tutorials seems to appear under the build folder. Even if I run it off of my main branch (which is even with this one's) the problem still occurs.

Comment (iancze) [reply docs/tutorials/initializedirtyimage.py:77]: Is it possible this is a local pip install issue, as per this [discussion](https://github.com/MPoL-dev/MPoL/discussions/66#discussioncomment-814423)?

Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:77]: I don't think so because with pip list MPoL appears the way you said it should and both Sphinx and jupytext are also there. I have checked my local branch and I also do have that commit and gridded visibilities is _grid_visibilities. I do get some other errors when doing make html (like it says [Errno 2] No such file or directory: '_static/baselines/build/baselines.csv'.) but none of them appear the error about the 'nonexistent' tutorials.

Comment (iancze) [reply docs/tutorials/initializedirtyimage.py:77]: What happens if you do

```
$ python
>>> import mpol 
>>> mpol.__version__
>>> from mpol.gridding import Gridder
>>> dir(Gridder)
```
?

This should show the updated routines, like 

```['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_fliplr_cube', '_grid_visibilities', '_grid_weights', '_histogram_channel', '_histogram_cube', '_null_dirty_beam', 'get_dirty_beam', 'get_dirty_beam_area', 'get_dirty_image', 'sky_vis_gridded', 'to_pytorch_dataset']```

Also, are there any build errors that happen before the one you listed?

You might also try doing a ``make clean`` before you restart ``make html``, just in case stale files are messing something up.

Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:77]: All the updated routines appear and there are no build errors that happen before the one I listed. I have been doing the make clean to make html as well. I've also done fresh installs of it all and still get the same problem. I'll see what happens when I switch what computer I'm working on, hopefully that will work.

File: docs/tutorials/initializedirtyimage.py (hunk 7 above)
Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:171]: I redo the uninitialized parameters to show the difference between them and then reloading them, though displaying that isn't too important. Though I suppose I should change the comments to make that clearer.

Comment (iancze) [reply docs/tutorials/initializedirtyimage.py:171]: Yup, that makes sense.

File: docs/tutorials/initializedirtyimage.py (hunk 1 above)
Comment (RCF42) [reply docs/tutorials/initializedirtyimage.py:80]: I've tried it completely fresh from a new computer and it still didn't work so I'm mildly at a lost for why the problem's happening.

Comment (hgrzy) [reply docs/tutorials/initializedirtyimage.py:80]: In your directory and virtual machine if you do the equivalent of "pip show jinja2" for your system, do you get version 3.0.1 or version 2.11?

Comment (iancze) [reply docs/tutorials/initializedirtyimage.py:80]: After talking with Robert, I think his main issue stems from how `make` works (or doesn't) on windows. But I did notice this error when I was installing a fork recently. The strange thing is that it appears to be transient... I think it's nbsphinx that relies upon jinja2 2.11, but sometimes the latest jinja2 ends up being installed by accident. I'm not really sure how this is possible with pip, that the dependencies are resolved non-deterministically. @hgrzy have you come across this error too, and if so, was it semi-random?

Comment (hgrzy) [reply docs/tutorials/initializedirtyimage.py:80]: Yes, I was getting that error a week or two ago, but I just tried replicating it and it seems to install the jinja2 version that is desired even though in both cases I was using pip version 20.2.3. I saved the terminal as a text file when I was getting the error though and it always occurred in the the process of installing the developer package at the point where it's running the setup.py files, though it isn't any particular setup.py file. It also included this in the output:

`ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.`

Comment (hgrzy) [reply docs/tutorials/initializedirtyimage.py:80]: I just looked at the release history for nbsphinx. As of 10 hours ago pip install nbsphinx now installs version 0.8.6 (which seems to run jinja2 version 3.0.1 fine) instead of 0.8.5 (requires jinja2 version 2.11.0) which is probably why I cannot replicate the error. So maybe this is no longer an issue?

Comment (iancze) [reply docs/tutorials/initializedirtyimage.py:80]: Good detective work! Hopefully this is resolved for future builds---it seems specifically an nbsphinx problem. If it crops up again we'll need to pin the jinja2 version, which I'd like to avoid doing unnecessarily.
